File: src/Class/Class.r
#ifndef CLASS_H
#define CLASS_H

#include "Constructable.r"
#include "Destructable.r"
#include "InvocationStatus.r"
#include "Legacy_AllocationTable.r"
#include "Legacy_AllocationTableList.r"
#include "ObjectBase.r"
#include <stdio.h>
#include <stdlib.h>

// Forward declaration of incomplete type
typedef struct class Class;

struct class {

    /// `Class` implements `ObjectBase`.
    ObjectBase *thisObjectBase;

    /// Singleton for the whole class. Sensitive data. DO NOT TOUCH!
    Legacy_AllocationTable *CLASS_ALLOCATION_TABLE;

    /// Sensitive data. DO NOT TOUCH!
    void *allocatedAddress;

    enum InvocationStatus destructorInvocationStatus;

    enum InvocationStatus deleteFromAllocationTableInvocationStatus;

    int x;

    /// Printing method.
    void (*print)();

    /// A logic method.
    void (*addOneToX)();
};

Class *
ClassConstructorWithClassName(void *objectToSaveItsAddressToAllocationTable,
                              const char *className);

/**
 * @brief Creates a `new` "heap-allocated" instance of `Class` and initializes
 *        it.
 *
 * @attention Remember to `free` the returned value of this function.
 *
 * For example:
 * @code
 *
 * // Construct.
 * Class *class = ClassConstructor();
 *
 * // Do something...
 * (*class->print)(class);
 * (*class->addOneToX)(class);
 * (*class->print)(class);
 *
 * // Remember to `free`.
 * free(class);
 * @endcode
 *
 * @return A pointer to a `new` "heap-allocated" instance of `Class`.
 */
Class *ClassConstructor();

void *ClassDestructor(Class *class);

#endif //CLASS_H

File: src/Class/Class.c
#include "Class.r"

void print(Class *class) { printf("x = %d\n", class->x); }

void addOneToX(Class *class) { class->x += 1; }

void *ClassDestructor(Class *class) {
    if (class == NULL) { return NULL; }

    class->CLASS_ALLOCATION_TABLE->allocationAddressList
            ->deleteNodeThatHasTheGivenData(
                    class->CLASS_ALLOCATION_TABLE->allocationAddressList,
                    class);

    free(class->thisObjectBase);

    // Destruct self.
    free(class);

    return NULL;
}

void constructor_Class_fields(Class *class) {
    class->thisObjectBase = ObjectBaseConstructor();

    static Constructable const constructable = {
            .constructor = (void *(*const)(void) )(&ClassConstructor)};
    class->thisObjectBase->constructable = &constructable;

    static Destructable const destructable = {
            .destructor = (void *(*const)(void *) )(&ClassDestructor)};
    class->thisObjectBase->destructable = &destructable;

    class->x         = 1;
    class->print     = &print;
    class->addOneToX = &addOneToX;
}

Class *ClassConstructorWithClassName(const char *className) {
    Class *obj = calloc(1, sizeof *obj);
    if (obj == NULL) { /* error handling here */
    }

    constructor_Class_fields(obj);

    obj->thisObjectBase->CLASS_NAME = className;
    obj->CLASS_ALLOCATION_TABLE =
            getLegacy_AllocationTableList()
                    ->findLegacy_AllocationTableByClassName(
                            obj->thisObjectBase->CLASS_NAME);
    if (obj->CLASS_ALLOCATION_TABLE == NULL) {
        obj->CLASS_ALLOCATION_TABLE =
                Legacy_AllocationTableConstructorWithClassName(
                        (char *) obj->thisObjectBase->CLASS_NAME);

        // Create a legacy_node that its data points to `obj->CLASS_ALLOCATION_TABLE`.
        Legacy_Node *nodeThatItsDataPointsClassAllocationTable =
                Legacy_NodeConstructorWithDataAndDataSize(
                        obj->CLASS_ALLOCATION_TABLE,
                        sizeof(Legacy_AllocationTable *));

        // Add this legacy_node to `GLOBAL_ALLOCATION_TABLE_LIST->legacy_allocationTableList`.
        getLegacy_AllocationTableList()->allocationTableList->add(
                getLegacy_AllocationTableList()->allocationTableList,
                nodeThatItsDataPointsClassAllocationTable);
    }

    // Create a legacy_node that its data points to the "pointer of `obj`".
    Legacy_Node *nodeThatItsDataPointsToThePointerOfObj =
            Legacy_NodeConstructorWithDataAndDataSize(obj, sizeof(void *));

    // Add this legacy_node to `obj->CLASS_ALLOCATION_TABLE->allocationAddressList`.
    obj->CLASS_ALLOCATION_TABLE->allocationAddressList->add(
            obj->CLASS_ALLOCATION_TABLE->allocationAddressList,
            nodeThatItsDataPointsToThePointerOfObj);

    return obj;
}

Class *ClassConstructor() { return ClassConstructorWithClassName("Class"); }
